### PR TITLE
Add Symfony 5 compatibility

### LINKED ISSUE #5788: Symfony 5 compatibility
I am trying to install sonata admin bundle in a fresh Symfony 5 installation but I get this error :
Problem 1
    - Conclusion: don't install sonata-project/admin-bundle 3.56.1
    - Conclusion: remove twig/twig v3.0.0
    - Installation request for sonata-project/admin-bundle ^3.56 -> satisfiable by sonata-project/admin-bundle[3.56.0, 3.56.1].
    - Conclusion: don't install twig/twig v3.0.0
    - sonata-project/admin-bundle 3.56.0 requires twig/twig ^2.10 -> satisfiable by twig/twig[v2.10.0, v2.11.0, v2.11.1, v2.11.2, v2.11.3, v2.12.0, v2.12.1, v2.12.2].
    - Can only install one of: twig/twig[v2.10.0, v3.0.0].
    - Can only install one of: twig/twig[v2.11.0, v3.0.0].
    - Can only install one of: twig/twig[v2.11.1, v3.0.0].
    - Can only install one of: twig/twig[v2.11.2, v3.0.0].
    - Can only install one of: twig/twig[v2.11.3, v3.0.0].
    - Can only install one of: twig/twig[v2.12.0, v3.0.0].
    - Can only install one of: twig/twig[v2.12.1, v3.0.0].
    - Can only install one of: twig/twig[v2.12.2, v3.0.0].
    - Installation request for twig/twig (locked at v3.0.0) -> satisfiable by twig/twig[v3.0.0].

### PR DESCRIPTION
**Pull request by franmomu**
## Subject

~Since https://github.com/sonata-project/SonataAdminBundle/pull/6212, KnpMenuBundle is the only external library that is not  compatible with Symfony 5 on the stable branch.~

~So I used https://github.com/sonata-project/SonataAdminBundle/pull/5948 that modifies classes that are marked as final.~

~The other BC-break that I introduced (that I'm aware of) is to add a needed return type `iterable` to `getExtendedTypes()` method in the classes (marked as final also) that implement `FormTypeExtensionInterface`.~

I copied `ControllerTrait` from Symfony since it's not available in Symfony 5.

I'm aware of https://github.com/sonata-project/SonataAdminBundle/pull/6093 and this is complementary since looks like the main problem for `4.0` is with `CRUDController`, so that still needs to be solved.

So the question is if this can be acceptable in the stable branch or not, if so we should also consider applying similar to https://github.com/sonata-project/SonataAdminBundle/pull/6212 (adding support to BlockBundle 4) to the bundles we want to be compatible with Symfony 5.

I am targeting this branch, because it would be nice to have Symfony 5 compatibility on the stable branch.

Closes #5788.

## Changelog

```markdown
### Added
- Added support for Symfony 5.
```

## To do

- [x] Release SonataIntlBundle.
- [ ] Fix PHPStan.
- [ ] Update the documentation.
- [ ] Add an upgrade note.

### CODE REVIEW

**Comment by vladyslavstartsev** [comment]
> The other BC-break that I introduced (that I'm aware of) is to add a needed return type iterable to getExtendedTypes()

what you can do (which does not break BC, but looks ugly)
is something like this

```php


if ($symfonyVersion === 4) {

class ChoiceTypeExtension extends AbstractTypeExtension
    {
        public static function getExtendedTypes()
        {
            return [ChoiceType::class];
        }
    }
} else {
class ChoiceTypeExtension extends AbstractTypeExtension
    {
        public static function getExtendedTypes(): iterable
        {
            return [ChoiceType::class];
        }
    }
}
```

this will be a workaround for the symfony 5. And later SonataAdmin 4.x will be released

**Comment by vladyslavstartsev** [comment]
why don't you do the same `if/else` (ugly) thing with `KnpMenuBundle`?

I think `if/else` is more soft, than breaking things the hard way.

With it (`if/else`) I'll at least will not break my app if I don't extend classes with BC break.

**Comment by franmomu** [comment]
> why don't you do the same `if/else` (ugly) thing with `KnpMenuBundle`?
> 
> I think `if/else` is more soft, than breaking things the hard way.
> 
> With it (`if/else`) I'll at least will not break my app if I don't extend classes with BC break.

I think this case is different, 

In this case we are not supporting `2.x` version of `KnpMenuBundle`, just `3.x` so 

> why don't you do the same `if/else` (ugly) thing with `KnpMenuBundle`?
> 
> I think `if/else` is more soft, than breaking things the hard way.
> 
> With it (`if/else`) I'll at least will not break my app if I don't extend classes with BC break.

I created https://github.com/sonata-project/SonataAdminBundle/pull/6275 where we can discuss about that and leave this to Symfony.

**Comment by jaikdean** [comment]
Looks like this is unblocked now. How can I help?

File: composer.json [hunk 1]
```diff
@@ -26,35 +26,36 @@
...
         "sonata-project/exporter": "^1.11 || ^2.0",
         "sonata-project/form-extensions": "^0.1.1 || ^1.4",
         "sonata-project/twig-extensions": "^0.1.1 || ^1.3",
-        "symfony/asset": "^4.4",
-        "symfony/config": "^4.4",
-        "symfony/console": "^4.4",
-        "symfony/dependency-injection": "^4.4.3",
-        "symfony/doctrine-bridge": "^4.4",
-        "symfony/event-dispatcher": "^4.4",
+        "symfony/asset": "^4.4 || ^5.1",
```
Comment (jaikdean) [inline composer.json:35]: What's the reason for `^4.4 || ^5.1` on Symfony packages rather than `^4.4 || ^5.0`? Is there something hard to support in 5.0?

Comment (dmaicher) [reply composer.json:35]: Symfony 5.0 is EOL (not supported anymore). See https://symfony.com/releases

So does not really make sense to support it here I believe :blush:

File: .travis.yml [hunk 2]
```diff
@@ -34,6 +34,8 @@ matrix:
       env: COMPOSER_FLAGS="--prefer-lowest"
     - php: '7.3'
       env: SYMFONY=4.4.*
+    - php: '7.3'
+      env: SYMFONY=5.1.*
     - php: '7.3'
       env: SYMFONY='dev-master as 4.4.x-dev'
```
Comment (jaikdean) [inline .travis.yml:40]: Should we update these environment variables from `dev-master as 4.4.x-dev` to `dev-master as 5.1.x-dev`? It should ensure other dependencies are used at the most likely versions to work?

File: composer.json (hunk 1 above)
Comment (jaikdean) [reply composer.json:35]: So it is, sorry 🤦

**Comment by wbloszyk** [comment]
@franmomu 
Could you move this PR to master branch? We can't add support for Symfony5 in Sonata3 becouse [Symfony 5 - TODO list](https://github.com/sonata-project/SonataAdminBundle/issues/6365)

**Comment by jaikdean** [comment]
> @franmomu
> Could you move this PR to master branch? We can't add support for Symfony5 in Sonata3 becouse [Symfony 5 - TODO list](https://github.com/sonata-project/SonataAdminBundle/issues/6365)

What BC breaks are in this PR? I've seen reference to BC breaks across a few issues/PRs, but not any specific examples.

File: src/Form/Extension/ChoiceTypeExtension.php [hunk 3]
```diff
@@ -13,38 +13,36 @@
...
-    {
-        $optionalOptions = ['sortable'];
-
-        $resolver->setDefined($optionalOptions);
-    }
-
-    public function buildView(FormView $view, FormInterface $form, array $options)
+use Symfony\Component\Form\FormTypeExtensionInterface;
+
+// NEXT_MAJOR: Remove the "else" part, copy all methods from BaseChoiceTypeExtension in this class and
```
Comment (jordisala1991) [inline src/Form/Extension/ChoiceTypeExtension.php:20]: IMHO this is a no go for maintenance cost...

**Comment by wbloszyk** [comment]
@jaikdean 
```json
        "symfony/framework-bundle": "^4.4 || ^5.1",
```
Using `^5.1` will not be possible - `class CoreController extends Controller`, but `Symfony\Bundle\FrameworkBundle\Controller\Controller` don't exists in Symfony5.

**Comment by wbloszyk** [comment]
> > @jaikdean
> > ```json
> >         "symfony/framework-bundle": "^4.4 || ^5.1",
> > ```
> > 
> > 
> > Using `^5.1` will not be possible - `class CoreController extends Controller`, but `Symfony\Bundle\FrameworkBundle\Controller\Controller` don't exists in Symfony5.
> 
> `CoreController` has been refactored in this PR to no longer extend `Symfony\Bundle\FrameworkBundle\Controller\Controller`.

It is BC-break.

What about translator? There are huge BC-break too.
https://github.com/sonata-project/SonataAdminBundle/pull/6342

IMO better will be simply add support for Symfony5 in Sonata4, without BC-break. This will save time which we can spend on major release. This move will also encourage people to upgrade Admin to v4 and make less work for as to support v3.

**Comment by alytvynov** [comment]
Hello! 
I suddenly found that sonata not compatible with `symfony/console` ^5.0.
Here in packaginst https://packagist.org/packages/sonata-project/admin-bundle sonata needs symfony/console: ^4.4. Anybody know something about this issue ?

**Comment by VincentLanglet** [comment]
> Hello!
> I suddenly found that sonata not compatible with `symfony/console` ^5.0.
> Here in packaginst https://packagist.org/packages/sonata-project/admin-bundle sonata needs symfony/console: ^4.4. Anybody know something about this issue ?

See https://github.com/sonata-project/SonataAdminBundle/issues/6365
Sonata 3.x won't be compatible with Symfony 5. We working on Sonata 4.

**Comment by alytvynov** [comment]
Thank you for your answer.

**Comment by VincentLanglet** [comment]
Closing in favor of https://github.com/sonata-project/SonataAdminBundle/pull/6476

**Pull request closed without merging**